src/transforms/pipelines/config.rs

```diff
@@ -42,62 +52,178 @@ impl Clone for PipelineConfig {
     }
 }
 
+#[async_trait::async_trait]
+#[typetag::serde(name = "pipeline")]
+impl TransformConfig for PipelineConfig {
+    async fn build(&self, ctx: &TransformContext) -> crate::Result<Transform> {
+        let condition = match &self.filter {
+            Some(config) => Some(config.build(&ctx.enrichment_tables)?),
+            None => None,
+        };
+
+        // Setup the interior transform chain in this pipeline.
+        //
+        // NOTE: As this pipeline is not expanding to separate transforms we
+        // lose validation that happens at the topology layer (e.g. making sure
+        // input/output types of connected transforms are compatible). This is
+        // an unfortunate consequence of having a "topology in a topology".
+        //
+        // In the future, it would be great to decouple that type of validation
+        // from the physical task boundaries. You could imagine a future where
+        // we replace transform expansion with the ability to return different
+        // types of compound transforms directly from TransformConfig::build,
+        // and those know how to play nice with a subsequent graph-based
+        // validation scheme.
+        //
+        // For now, considering that most everything is log to log with regard
+        // to pipelines we avoid adding duplicate validation here in favor of
+        // future work.
+        if self.transforms.is_empty() {
+            return Err(format!("empty pipeline: {}", self.name).into());
+        }
+        // Today we make the assumption that to be a valid pipeline transform
+        // the transform CANNOT have named outputs. This assumption might break
+        // in the future so, to avoid panics, we instead make building a
+        // pipeline with such transforms an error.
+        for transform in &self.transforms {
+            if transform.outputs(&ctx.merged_schema_definition).len() > 1 {
+                return Err(format!(
+                    "pipeline {} has transform of type {} with a named output, unsupported",
+                    self.name,
+                    transform.transform_type()
+                )
+                .into());
+            }
+        }
+
+        let mut transforms = Vec::with_capacity(self.transforms.len());
+        for config in &self.transforms {
+            let transform = match config.build(ctx).await? {
+                Transform::Function(transform) => Box::new(transform),
+                Transform::Synchronous(transform) => transform,
+                _ => return Err(format!("non-sync transform in pipeline: {:?}", config).into()),
```

Comment:
I don't think we really expect users to ever encounter this error, since the pipeline transform isn't something they'll be configuring themselves (if I understand correctly). The stateful/stateless distinction is not accurate here so I would not use it. Maybe something more generic like "unexpected transform type within pipeline" or similar?

Reply:
I realize we are doing automatic translation too, but this `pipelines` feature will be generally available and usable by users so I would expect we'll see some configuring it themselves.

`unexpected transform type within pipeline` is better, but still doesn't tell the user any information to know which transform types _are_ allowed. Do we need to start documenting on the website which are function vs. synchronous vs. task? I'm still not sure I understand the distinction between function and synchronous.

Reply:
I'd be a little nervous about documenting these differences, at least that clearly. I'd like the chance to change things without needing to go through a deprecation cycle, which we might trip if we document too clearly. 

But, your point here is a valid one. Right now the error is just a big "NO" with no user valuable context.